// parcel-bundler/lib/assets/TypeScriptAsset.js
'use strict';

function _asyncToGenerator(fn) { return function () { var gen = fn.apply(this, arguments); return new Promise(function (resolve, reject) { function step(key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { return Promise.resolve(value).then(function (value) { step("next", value); }, function (err) { step("throw", err); }); } } return step("next"); }); }; }

const JSAsset = require('./JSAsset');
const localRequire = require('../utils/localRequire');

class TypeScriptAsset extends JSAsset {
  parse(code) {
    var _this = this;

    return _asyncToGenerator(function* () {
      // require typescript, installed locally in the app
      let typescript = yield localRequire('typescript', _this.name);
      let transpilerOptions = {
        compilerOptions: {
          module: typescript.ModuleKind.CommonJS,
          jsx: typescript.JsxEmit.Preserve
        },
        fileName: _this.relativeName
      };

      let tsconfig = yield _this.getConfig(['tsconfig.json']);

      // Overwrite default if config is found
      if (tsconfig) {
        transpilerOptions.compilerOptions = Object.assign(transpilerOptions.compilerOptions, tsconfig.compilerOptions);
      }
      transpilerOptions.compilerOptions.noEmit = false;
      transpilerOptions.compilerOptions.sourceMap = _this.options.sourceMaps;

      // Transpile Module using TypeScript and parse result as ast format through babylon
      let transpiled = typescript.transpileModule(code, transpilerOptions);
      _this.sourceMap = transpiled.sourceMapText;

      if (_this.sourceMap) {
        _this.sourceMap = JSON.parse(_this.sourceMap);
        _this.sourceMap.sources = [_this.relativeName];
        _this.sourceMap.sourcesContent = [_this.contents];

        // Remove the source map URL
        let content = transpiled.outputText;
        transpiled.outputText = content.substring(0, content.lastIndexOf('//# sourceMappingURL'));
      }

      _this.contents = transpiled.outputText;
      return yield JSAsset.prototype.parse.call(_this, _this.contents);
    })();
  }
}

module.exports = TypeScriptAsset;
